File: backend-node/lib/tracer.js
// Centralized event tracer — single place for all debug logs
const { decode } = require('@msgpack/msgpack')
const logger = require('./logger')

const INBOUND = '\u2190'
const OUTBOUND = '\u2192'

function kafkaProduce(topic, key, value) {
  const summary = summarize(value)
  logger.info({ target: 'kafka', topic, key, summary }, `KAFKA ${OUTBOUND} ${topic} key=${key} ${summary}`)
}

function kafkaConsume(topic, key, value) {
  const keyStr = key == null ? '' : String(key)
  const summary = summarize(value)
  logger.info({ target: 'kafka', topic, key: keyStr, summary }, `KAFKA ${INBOUND} ${topic} key=${keyStr} ${summary}`)
}

function connConnect(transport, userId, peer) {
  logger.info({ target: 'conn', transport, userId, peer }, `${transport.toUpperCase()} ${INBOUND} user=${userId} from ${peer}`)
}

function connDisconnect(transport, userId, peer) {
  logger.info({ target: 'conn', transport, userId, peer }, `${transport.toUpperCase()} ${OUTBOUND} user=${userId} disconnected from ${peer}`)
}

function msgReceived(transport, userId, msgType, size) {
  logger.info({ target: 'msg', transport, userId, msgType, size }, `${transport.toUpperCase()} ${INBOUND} user=${userId} type=${msgType} size=${size}`)
}

function msgSent(transport, userId, msgType, size) {
  logger.info({ target: 'msg', transport, userId, msgType, size }, `${transport.toUpperCase()} ${OUTBOUND} user=${userId} type=${msgType} size=${size}`)
}

function authEvent(transport, userId, peer, success) {
  if (success) {
    logger.info({ target: 'auth', transport, userId, peer }, `AUTH OK ${userId} ${peer} via ${transport}`)
  } else {
    logger.warn({ target: 'auth', transport, peer }, `AUTH FAIL from ${peer} via ${transport}`)
  }
}

function summarize(value) {
  if (!value) return '(empty)'
  try {
    const buf = Buffer.isBuffer(value) ? value : Buffer.from(value)
    const obj = decode(buf)
    if (typeof obj === 'object' && obj !== null) {
      const parts = []
      for (const key of ['chatId', 'userId', 'senderId', 'messageId', 'type', 'online']) {
        if (obj[key] !== undefined) parts.push(`${key}:${obj[key]}`)
      }
      if (parts.length === 0) {
        return '{' + Object.keys(obj).slice(0, 4).join(' ') + '}'
      }
      return '{' + parts.join(' ') + '}'
    }
    return String(obj)
  } catch {
    const len = Buffer.isBuffer(value) ? value.length : (value ? value.length || 0 : 0)
    return `raw[${len}b]`
  }
}

module.exports = { kafkaProduce, kafkaConsume, connConnect, connDisconnect, msgReceived, msgSent, authEvent }

File: backend-rust/src/tracer.rs
// Centralized event tracer — single place for all debug logs
// Format: [category] direction event_type key_fields

use tracing::{info, warn};

pub enum Direction {
    Inbound,
    Outbound,
}

impl Direction {
    fn arrow(&self) -> &'static str {
        match self {
            Direction::Inbound => "\u{2190}",
            Direction::Outbound => "\u{2192}",
        }
    }
}

pub fn kafka_produce(topic: &str, key: &str, payload: &[u8]) {
    let summary = summarize_msgpack(payload);
    info!(
        target: "kafka",
        topic = topic,
        key = key,
        summary = summary,
        "KAFKA {} {} key={} {}",
        Direction::Outbound.arrow(),
        topic,
        key,
        summary,
    );
}

pub fn kafka_consume(topic: &str, key: &[u8], payload: &[u8]) {
    let key_str = String::from_utf8_lossy(key);
    let summary = summarize_msgpack(payload);
    info!(
        target: "kafka",
        topic = topic,
        key = key_str.as_ref(),
        summary = summary,
        "KAFKA {} {} key={} {}",
        Direction::Inbound.arrow(),
        topic,
        key_str,
        summary,
    );
}

pub fn conn_connect(transport: &str, user_id: i64, peer: &str) {
    info!(
        target: "conn",
        transport = transport,
        user_id = user_id,
        peer = peer,
        "{} {} user={} from {}",
        transport.to_uppercase(),
        Direction::Inbound.arrow(),
        user_id,
        peer,
    );
}

pub fn conn_disconnect(transport: &str, user_id: i64, peer: &str) {
    info!(
        target: "conn",
        transport = transport,
        user_id = user_id,
        peer = peer,
        "{} {} user={} disconnected from {}",
        transport.to_uppercase(),
        Direction::Outbound.arrow(),
        user_id,
        peer,
    );
}

pub fn msg_received(transport: &str, user_id: i64, msg_type: u16, size: usize) {
    info!(
        target: "msg",
        transport = transport,
        user_id = user_id,
        msg_type = msg_type,
        size = size,
        "{} {} user={} type={} size={}",
        transport.to_uppercase(),
        Direction::Inbound.arrow(),
        user_id,
        msg_type,
        size,
    );
}

pub fn msg_sent(transport: &str, user_id: i64, msg_type: u16, size: usize) {
    info!(
        target: "msg",
        transport = transport,
        user_id = user_id,
        msg_type = msg_type,
        size = size,
        "{} {} user={} type={} size={}",
        transport.to_uppercase(),
        Direction::Outbound.arrow(),
        user_id,
        msg_type,
        size,
    );
}

pub fn auth_event(transport: &str, user_id: i64, peer: &str, success: bool) {
    if success {
        info!(
            target: "auth",
            transport = transport,
            user_id = user_id,
            peer = peer,
            "AUTH OK {} {} via {}",
            user_id,
            peer,
            transport,
        );
    } else {
        warn!(
            target: "auth",
            transport = transport,
            peer = peer,
            "AUTH FAIL from {} via {}",
            peer,
            transport,
        );
    }
}

fn summarize_msgpack(data: &[u8]) -> String {
    if data.is_empty() {
        return "(empty)".into();
    }
    match rmp_serde::from_slice::<serde_json::Value>(data) {
        Ok(val) => {
            let mut parts: Vec<String> = Vec::new();
            if let Some(obj) = val.as_object() {
                for key in ["chatId", "userId", "senderId", "messageId", "type", "online"] {
                    if let Some(v) = obj.get(key) {
                        parts.push(format!("{}:{}", key, v));
                    }
                }
                if parts.is_empty() {
                    let keys: Vec<&str> = obj.keys().map(|k| k.as_str()).take(4).collect();
                    parts = keys.iter().map(|k| format!("{}", k)).collect();
                }
            } else {
                return format!("{:?}", val);
            }
            if parts.is_empty() { "{...}".into() } else { format!("{{{}}}", parts.join(" ")) }
        }
        Err(_) => format!("raw[{}b]", data.len()),
    }
}
